mayastor/src/bdev/nexus/nexus_child.rs

```diff
@@ -33,40 +25,21 @@ pub enum ChildError {
     OpenChild { source: CoreError },
     #[snafu(display("Claim child"))]
     ClaimChild { source: Errno },
-    #[snafu(display("Child is read-only"))]
-    ChildReadOnly {},
+    #[snafu(display("Child is closed"))]
+    ChildClosed {},
     #[snafu(display("Invalid state of child"))]
     ChildInvalid {},
     #[snafu(display("Opening child bdev without bdev pointer"))]
     OpenWithoutBdev {},
     #[snafu(display("Failed to create a BdevHandle for child"))]
     HandleCreate { source: CoreError },
-
-    #[snafu(display("Failed to read MBR from child"))]
-    MbrRead { source: ChildIoError },
-    #[snafu(display("MBR is invalid"))]
-    MbrInvalid { source: LabelError },
-
-    #[snafu(display("Failed to allocate buffer for label"))]
-    LabelAlloc { source: DmaError },
-    #[snafu(display("Failed to read label from child"))]
-    LabelRead { source: ChildIoError },
-    #[snafu(display("Label is invalid"))]
-    LabelInvalid {},
-
-    #[snafu(display("Failed to allocate buffer for partition table"))]
-    PartitionTableAlloc { source: DmaError },
-    #[snafu(display("Failed to read partition table from child"))]
-    PartitionTableRead { source: ChildIoError },
-    #[snafu(display("Partition table is invalid"))]
-    PartitionTableInvalid { source: LabelError },
 }
 
 #[derive(Debug, Snafu)]
 pub enum ChildIoError {
-    #[snafu(display("Error writing to {}", name))]
+    #[snafu(display("Error writing to {}: {}", name, source))]
```

Comment:
Actually you do need the source here explicitly if you want its message to be included. And rust will definitely complain if the number of "{}" specifiers does not match the number of arguments. What is true is that you do not need to set the "source" field explicitly when you are creating a "context" for this error as snafu will automatically do that for you.

Reply:
Actually you don't strictly need to... what I meant was not adding the {} as well as the source.

The created error will indeed have the source information as the snafu ext does this for you. But extra information might not included in the default Display trait, you'd have to extend it. You need however to unwind the error chain when printing errors which we do on rpc/grpc code... yes it's a bit messy what happens when there are sources in the display message (not just because of your code) is that some rpc errors end up with duplicated messages, eg:
Error: RpcError { code: InternalError, msg: "Failed to write label of nexus nexus-be8d0ecf-ed31-497e-a554-56c47d401b4c: Write error: **Invalid descriptor for child bdev aio:///dev/ram0: Invalid descriptor for child bdev aio:///dev/ram0"** }

So we might need to remove the sources from all messages, or add them to all.  @gila  any thoughts?